### PR TITLE
HDDS-10518. Create base Builder in WithMetadata and WithObjectID

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Builders in subclasses of `WithMetadata` and `WithObjectID` all have their own code for setting the properties from the base classes.  The goal of this task is to reduce code duplication by creating builders in the base classes and using them in the subclasses.

Existing subclass builder methods are kept for covariant return to be able to keep existing chained method calls (e.g. return `OmBucketInfo` instead of `WithMetadata`).  We can remove them after replacing with non-chained calls in a follow-up patch.

https://issues.apache.org/jira/browse/HDDS-10518

## How was this patch tested?

CI:
https://github.com/adoroszlai/ozone/actions/runs/8277449313

### CODE REVIEW

**Comment by adoroszlai** [comment]
@ivandika3 please take a look as well

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java [hunk 1]
```diff
@@ -984,7 +983,7 @@ private Builder(ConfigurationSource conf, ClientProtocol proxy) {
     }
 
     public Builder setMetadata(Map<String, String> metadata) {
-      this.metadata = metadata;
+      addAllMetadata(metadata);
```
Comment (ivandika3) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java:986]: Nit: Should not affect anything, but the semantic seems to be changed changed from "replacing" to "adding". Might need to clear the metadata before adding it or can change the method name to `addAllMetadata` (but `OzoneBucket` is client-facing class so I don't suggest this).

I'm OK if you decide not to change it at all.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java [hunk 2]
```diff
@@ -483,7 +482,7 @@ public Builder setRefCount(long refCount) {
     }
 
     public Builder setMetadata(Map<String, String> metadata) {
-      this.metadata = metadata;
+      addAllMetadata(metadata);
```
Comment (ivandika3) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneVolume.java:485]: Similar comment to `OzoneBucket.Builder#setMetadata`.

**Review by ivandika3 — COMMENTED** [review]
Thank you for the patch. Mostly LGTM. Few minor comments.

**Review by ivandika3 — APPROVED** [review]
Thank you for the update. LGTM +1.

**Review by szetszwo — COMMENTED** [review]
@adoroszlai , thanks for working on this!  The change look good.  I just have comments on adding `@Override` tag.

**Comment by adoroszlai** [comment]
Thanks @szetszwo for the review, added all missing `@Override` annotations.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by adoroszlai** [comment]
Thanks @ivandika3, @szetszwo for the review.

**Pull request merged**